subxt/src/client/lightclient/background.rs

```diff
@@ -0,0 +1,428 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use futures::stream::StreamExt;
+use futures_util::future::{self, Either};
+use serde::Deserialize;
+use serde_json::value::RawValue;
+use std::{collections::HashMap, str::FromStr};
+use tokio::sync::{mpsc, oneshot};
+
+use super::LightClientError;
+use smoldot_light::{platform::async_std::AsyncStdTcpWebSocket as Platform, ChainId};
+
+const LOG_TARGET: &str = "light-client-background";
+
+/// The response of an RPC method.
+pub type MethodResponse = Result<Box<RawValue>, LightClientError>;
+
+/// Message protocol between the front-end client that submits the RPC requests
+/// and the backend handler that produces responses from the chain.
+///
+/// The light client uses a single object [`smoldot_light::JsonRpcResponses`] to
+/// handle all requests and subscriptions from a chain. A background task is spawned
+/// to multiplex the rpc responses and to provide them back to their rightful submitters.
+#[derive(Debug)]
+pub enum FromSubxt {
+    /// The RPC method request.
+    Request {
+        /// The method of the request.
+        method: String,
+        /// The parameters of the request.
+        params: String,
+        /// Channel used to send back the result.
+        sender: oneshot::Sender<MethodResponse>,
+    },
+    /// The RPC subscription (pub/sub) request.
+    Subscription {
+        /// The method of the request.
+        method: String,
+        /// The parameters of the request.
+        params: String,
+        /// Channel used to send back the subscription ID if successful.
+        sub_id: oneshot::Sender<MethodResponse>,
+        /// Channel used to send back the notifcations.
+        sender: mpsc::Sender<Box<RawValue>>,
+    },
+}
+
+/// Background task data.
+pub struct BackgroundTask {
+    /// Smoldot light client implementation that leverages the exposed platform.
+    client: smoldot_light::Client<Platform>,
+    /// The ID of the chain used to identify the chain protocol (ie. substrate).
+    ///
+    /// Note: A single chain is supported for a client. This aligns with the subxt's
+    /// vision of the Client.
+    chain_id: ChainId,
+    /// Unique ID for RPC calls.
+    request_id: usize,
+    /// Map the request ID of a RPC method to the frontend `Sender`.
+    requests: HashMap<String, oneshot::Sender<MethodResponse>>,
```

Comment:
why are request IDs stored as Strings instead of usize? 

I prefer usize here and avoid a needless allocation :)

Reply:
It's something I wasn't entirely sure too. I was basically afraid that the subscription ID might be handed back in hex format since it's actually a number inside a string (ie result: "123").
But I think for now we could go away without this allocation and revisit things if subscription ID breaks compatibility over time